rust-sdk/core/src/quote/swap.rs

```diff
@@ -0,0 +1,653 @@
+                .ok_or(ARITHMETIC_OVERFLOW)?;
+            amount_calculated = amount_calculated
+                .checked_add(step_quote.amount_out)
+                .ok_or(ARITHMETIC_OVERFLOW)?;
+        } else {
+            amount_remaining = amount_remaining
+                .checked_sub(step_quote.amount_out)
+                .ok_or(ARITHMETIC_OVERFLOW)?;
+            amount_calculated = amount_calculated
+                .checked_add(step_quote.amount_in)
+                .ok_or(ARITHMETIC_OVERFLOW)?
+                .checked_add(step_quote.fee_amount)
+                .ok_or(ARITHMETIC_OVERFLOW)?;
+        }
+
+        if step_quote.next_sqrt_price == next_tick_sqrt_price {
+            current_liquidity = get_next_liquidity(current_liquidity, next_tick, a_to_b);
+            current_tick_index = if a_to_b {
+                next_tick_index - 1
+            } else {
+                next_tick_index
+            }
+        } else {
+            current_tick_index = sqrt_price_to_tick_index(step_quote.next_sqrt_price.into()).into();
+        }
+
+        current_sqrt_price = step_quote.next_sqrt_price;
+    }
+
+    let swapped_amount = token_amount - amount_remaining;
+
+    let token_a = if a_to_b == specified_input {
+        swapped_amount
+    } else {
+        amount_calculated
+    };
+    let token_b = if a_to_b == specified_input {
+        amount_calculated
+    } else {
+        swapped_amount
+    };
+
+    Ok(SwapResult {
+        token_a,
+        token_b,
+        trade_fee,
+    })
+}
+
+fn get_next_liquidity(current_liquidity: u128, next_tick: &TickFacade, a_to_b: bool) -> u128 {
+    let liquidity_net = next_tick.liquidity_net;
+    let liquidity_net_unsigned = liquidity_net.unsigned_abs();
+    if a_to_b {
+        if liquidity_net < 0 {
+            current_liquidity + liquidity_net_unsigned
+        } else {
+            current_liquidity - liquidity_net_unsigned
+        }
+    } else if liquidity_net < 0 {
+        current_liquidity - liquidity_net_unsigned
+    } else {
+        current_liquidity + liquidity_net_unsigned
+    }
+}
+
+struct SwapStepQuote {
+    amount_in: u64,
+    amount_out: u64,
+    next_sqrt_price: u128,
+    fee_amount: u64,
+}
+
+fn compute_swap_step(
+    amount_remaining: u64,
+    fee_rate: u16,
+    current_liquidity: u128,
+    current_sqrt_price: u128,
+    target_sqrt_price: u128,
+    a_to_b: bool,
+    specified_input: bool,
+) -> Result<SwapStepQuote, ErrorCode> {
+    let initial_amount_fixed_delta = try_get_amount_fixed_delta(
+        current_sqrt_price,
+        target_sqrt_price,
+        current_liquidity,
+        a_to_b,
+        specified_input,
+    );
+
+    let amount_calculated = if specified_input {
+        try_apply_swap_fee(amount_remaining.into(), fee_rate)?
+    } else {
+        amount_remaining
+    };
+
+    let next_sqrt_price =
+        if initial_amount_fixed_delta.is_ok() && initial_amount_fixed_delta? <= amount_calculated {
```

Comment:
hmm, could you use same logic to the contract ?
There are several advantages to doing so.

- Battle-tested code can be used.
- It reduces review time because it is the same code. (It is more effective to spend time on completely new code than to spend time checking that it produces the same results for different code.)
- If we decide to use `core` for contract, it is relatively easy to say `those are the same`.

In this part, `AmountDeltaU64` is introduced to distinguish unrecoverable error and u64 overflow issue.

https://github.com/orca-so/whirlpools/blob/main/programs/whirlpool/src/math/token_math.rs#L80
https://github.com/orca-so/whirlpools/pull/244 (`do not cover MultiplicationOverflow on try_get_amount_delta_a` section).

So `is_err()` has edge-case.